File: assets/classes/status.py
class Status:
    def __init__(self, hp, maxHp, regenHp, mana, maxMana, regenMana, stamina, maxStamina, regenStamina, damage, critical, defense, speed, ace):
        self.hp = hp
        self.maxHp = maxHp
        self.regenHp = regenHp
        self.mana = mana
        self.maxMana = maxMana
        self.regenMana = regenMana
        self.stamina = stamina
        self.maxStamina = maxStamina
        self.regenStamina = regenStamina
        self.damage = damage 
        self.critical = critical
        self.defense = defense
        self.speed = speed
        self.ace = ace
        self.active_effects = []

    def buff(self, stat, value):
        if hasattr(self, stat):  # Verifica se o atributo existe
            setattr(self, stat, getattr(self, stat) + value)  # Atualiza o valor do atributo
    
    def get(self, stat):
        if hasattr(self, stat):
            return getattr(self, stat)

    def set(self, stat, value):
        if hasattr(self, stat):
            setattr(self, stat, value)

    def add_effect(self, buff):
        # Aplica o efeito imediatamente
        setattr(self, buff.stat, getattr(self, buff.stat) + buff.value)
        buff.applied = True
        self.active_effects.append(buff)

    def update_effects(self):
        for buff in self.active_effects[:]:
            buff.remaining -= 1
            if buff.remaining <= 0:
                # Remove o efeito
                if buff.applied:
                    setattr(self, buff.stat, getattr(self, buff.stat) - buff.value)
                self.active_effects.remove(buff)
class Effect:
    def __init__(self, stat, value, duration):
        self.stat = stat
        self.value = value
        self.duration = duration  # Em ticks
        self.remaining = duration
        self.applied = False

File: server/models/Status.js
class Status{
    constructor(hp,maxHp,hpRegen,maxMana,mana,manaRegen,stamina,maxStamina,staminaRegen,damage,critical,defense,speed,aceleration){
        this.maxHp = maxHp;
        this.hp = hp;
        this.hpRegen = hpRegen;
        this.maxMana = maxMana;
        this.mana = mana;
        this.manaRegen = manaRegen;
        this.maxStamina = maxStamina;
        this.stamina = stamina;
        this.staminaRegen = staminaRegen;
        this.damage = damage;
        this.critical = critical;
        this.defense = defense;
        this.speed = speed;
        this.aceleration = aceleration;
        this.staminaRegen = staminaRegen;
        this.active_effects = [];
    }
    buff(stat,value){
        if(this[stat] !== undefined){
            this[stat] += value;
        }
    }
    get(stat){
        if(this[stat] !== undefined){
            return this[stat];
        } else {
            return null;
        }
    }
    set(stat,value){
        if(this[stat] !== undefined){
            this[stat] == value;
        }
    }
    add_effect(effect){
        this[effect.stat] = (this[effect.stat] || 0) + effect.value;
        effect.applied = true;
        this.active_effects.push(effect);
    }
    update_effects(){
        for(let effect of this.active_effects){
            if(effect.duration){
                effect.duration--;
            }
            if(effect.duration <= 0){
                if(effect.applied){
                    this[effect.stat] -= effect.value;
                }
                this.active_effects = this.active_effects.filter(e => e !== effect);
            }
        }
    }
}
class Effect{
    constructor(stat,value,duration){
        this.stat = stat;
        this.value = value;
        this.duration = duration;
        this.applied = false;
    }

}
export { Status, Effect };